src/Allocator/Allocator.hs

```diff
@@ -0,0 +1,27 @@
+module Allocator.Allocator where
+
+import Allocator.Process
+import Foundation
+import Control.Effect.Error
+import Control.Effect.State
+import Data.HashMap (Map)
+
+-- map of process id to its pages
+type PMap = Map Pid Mem
+
+data AllocError = OutOfMemory deriving (Eq, Ord)
+
+type AllocSig sig m = ( Has (State PMap) sig m
+                      , Has (Throw AllocError) sig m
+                      , Has (Catch AllocError) sig m)
+
+
+class Allocator a where
```

Comment:
We need to create another type for pointers(and we need to settle on logic that we will use for creating the) and return them here instead of `Word8`. Also, we need to hide fact that we passing `Process` instances inside this functions and expose the same API as in the C malloc library.